src/node.rs

```diff
@@ -0,0 +1,223 @@
+use crate::message::{FindValueResult, Message};
+use crate::routing_table::RoutingTable;
+use crate::utils::NodeId;
+use log::{debug, error, info, warn};
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha256};
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::net::{ UdpSocket};
+use std::net::SocketAddr;
+
+use tokio::sync::mpsc;
+use tokio::time::{interval, Duration};
+
+use crate::utils::{ALPHA, BOOTSTRAP_NODES, K};
+use bincode::{deserialize, serialize};
+
+pub struct KademliaNode {
+    id: NodeId,
+    addr: SocketAddr,
+    routing_table: RoutingTable,
+    storage: HashMap<Vec<u8>, Vec<u8>>,
+    socket: Arc<UdpSocket>,
+    shutdown: mpsc::Receiver<()>,
+}
+
+impl KademliaNode {
+    pub async fn new(addr: SocketAddr) -> std::io::Result<(Self, mpsc::Sender<()>)> {
+        let id = NodeId::new();
+        let socket = UdpSocket::bind(addr).await?;
+        let (shutdown_sender, shutdown_receiver) = mpsc::channel(1);
+
+        Ok((
+            KademliaNode {
+                id: id.clone(),
+                addr,
+                routing_table: RoutingTable::new(id),
+                storage: HashMap::new(),
+                socket: Arc::new(socket),
+                shutdown: shutdown_receiver,
+            },
+            shutdown_sender,
+        ))
+    }
+
+    pub async fn bootstrap(&mut self) -> std::io::Result<()> {
+        for &bootstrap_addr in BOOTSTRAP_NODES.iter() {
+            match bootstrap_addr.parse() {
+                Ok(addr) => {
+                    if let Err(e) = self.ping(addr).await {
+                        error!(
+                            "Failed to ping bootstrap node {:#?}: {:?}",
+                            bootstrap_addr, e
+                        );
+                    }
+                }
+                Err(e) => error!("Invalid bootstrap address {:#?}: {:?}", bootstrap_addr, e),
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn run(&mut self) -> std::io::Result<()> {
+        let mut buf = vec![0u8; 1024];
+        let mut refresh_interval = interval(Duration::from_secs(3600)); // Refresh every hour
```

Comment:
You might want to set [MissedTickBehavior](https://docs.rs/tokio/latest/tokio/time/enum.MissedTickBehavior.html) 😉